feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,567 @@
+			if pccRule.RefUmData == nil || len((*pccRule.RefUmData)[0]) == 0 {
+				continue
+			}
+			umEntry, found := smPolicyDecision.UmDecs.Get((*pccRule.RefUmData)[0])
+			if !found {
+				continue
+			}
+			monitors = append(monitors, &protos.UsageMonitoringUpdateResponse{
+				Credit:           GetUsageMonitoringCreditsProto((*pccRule.RefUmData)[0], &umEntry, protos.MonitoringLevel_PCC_RULE_LEVEL),
+				SessionId:        request.SessionId,
+				TgppCtx:          tgppCtx,
+				Sid:              request.GetCommonContext().GetSid().Id,
+				Success:          true,
+				EventTriggers:    eventTriggers,
+				RevalidationTime: revalidationTime,
+			})
+		}
+	}
+	// Usage monitoring at session level
+	if smPolicyDecision.SessRules != nil {
+		for _, sessRule := range smPolicyDecision.SessRules.AdditionalProperties {
+			if sessRule.RefUmData == nil || len(*sessRule.RefUmData) == 0 {
+				continue
+			}
+			umEntry, found := smPolicyDecision.UmDecs.Get(*sessRule.RefUmData)
+			if !found {
+				continue
+			}
+			monitors = append(monitors, &protos.UsageMonitoringUpdateResponse{
+				Credit:           GetUsageMonitoringCreditsProto(*sessRule.RefUmData, &umEntry, protos.MonitoringLevel_SESSION_LEVEL),
+				SessionId:        request.SessionId,
+				TgppCtx:          tgppCtx,
+				Sid:              request.GetCommonContext().GetSid().Id,
+				Success:          true,
+				EventTriggers:    eventTriggers,
+				RevalidationTime: revalidationTime,
+			})
+		}
+	}
+
+	return monitors
+}
+
+func GetUsageMonitoringCreditsProto(
+	umId string,
+	umData *n7_sbi.UsageMonitoringData,
+	monLevel protos.MonitoringLevel,
+) *protos.UsageMonitoringCredit {
+	monAction := protos.UsageMonitoringCredit_CONTINUE
+	if len(umData.UmId) == 0 {
+		// UmData not present for ID. It is supposed to be removed
+		monAction = protos.UsageMonitoringCredit_DISABLE
+	}
+	return &protos.UsageMonitoringCredit{
+		MonitoringKey: []byte(umId),
+		Action:        monAction,
+		Level:         monLevel,
+		GrantedUnits:  GetGrantedUnitsProto(umData),
+	}
+}
+
+func GetGrantedUnitsProto(umData *n7_sbi.UsageMonitoringData) *protos.GrantedUnits {
+	if umData == nil {
+		return &protos.GrantedUnits{
+			Total: &protos.CreditUnit{IsValid: false},
+			Tx:    &protos.CreditUnit{IsValid: false},
+			Rx:    &protos.CreditUnit{IsValid: false},
+		}
+	}
+	gsuProto := &protos.GrantedUnits{
+		Total: GetCreditUnitsProto(umData.VolumeThreshold),
+		Tx:    GetCreditUnitsProto(umData.VolumeThresholdUplink),   // Input == Tx == Uplink
+		Rx:    GetCreditUnitsProto(umData.VolumeThresholdDownlink), // Output == Rx == Downlink
+	}
+
+	return gsuProto
+}
+
+func GetCreditUnitsProto(volume *common5g.VolumeRm) *protos.CreditUnit {
+	if volume == nil {
+		return &protos.CreditUnit{IsValid: false}
+	}
+	return &protos.CreditUnit{IsValid: true, Volume: uint64(*volume)}
+}
+
+func ConvertToProtoTimeStamp(srcTime *time.Time) *timestamp.Timestamp {
+	if srcTime == nil {
+		return nil
+	}
+	return timestamppb.New(*srcTime)
+}
```

Comment:
please can you go over the file and group all the Proto fucntions and SBI functions toghetehr.  You can add a comment per each of section so it is visible


Another thing I personally consider good is to put all the small functions at the end (for example `BitRateStringToUint32` should be at the end of the proto block). The reason is those functions are kind of obvious, and they are distracting when reading the code.

Reply:
Done.